crest/Assets/Crest/Crest/Scripts/LodData/RegisterLodDataInput.cs

```diff
@@ -95,30 +130,113 @@ public static OceanInput GetRegistrar(Type lodDataMgrType)
         }
 
         internal Renderer _renderer;
-        protected Material _material;
+        protected Material _paintInputMaterial;
         // We pass this to GetSharedMaterials to avoid allocations.
         protected List<Material> _sharedMaterials = new List<Material>();
         SampleHeightHelper _sampleHelper = new SampleHeightHelper();
 
-        // If this is true, then the renderer should not be there as input source is from something else.
-        protected virtual bool RendererRequired => true;
         protected virtual bool SupportsMultiPassShaders => false;
 
-        void InitRendererAndMaterial(bool verifyShader)
+        protected virtual void OnEnable()
+        {
+        }
+
+        protected virtual void OnDisable()
+        {
+        }
+
+        public virtual bool AutoDetectMode(out InputMode mode)
         {
-            _renderer = GetComponent<Renderer>();
+            if (TryGetComponent<Renderer>(out _))
+            {
+                mode = InputMode.CustomGeometryAndShader;
+                return true;
+            }
+
+            mode = DefaultMode;
+            return false;
+        }
 
-            if (RendererRequired && _renderer != null)
+        protected virtual void Awake()
+        {
+            // Select a mode if it is unset. Running this in Awake() means it will
+            // run for components created at runtime ensuring they enter a valid state,
+            // and also helps data migration in edit mode (provided component is enabled).
+            if (_inputMode == InputMode.Unset)
             {
+                AutoDetectMode(out _inputMode);
+
+                // Make sure data change is registered if this runs in edit mode
 #if UNITY_EDITOR
-                if (Application.isPlaying && verifyShader)
+                EditorUtility.SetDirty(this);
```

Comment:
Calling _SetDirty_ here has a few issues sadly.

This will need to be wrapped to avoid always dirtying the main scene when ODC adds the SFD input:

```csharp
if (!Application.isPlaying && !hideFlags.HasFlag(HideFlags.DontSaveInEditor))
{
    EditorUtility.SetDirty(this);
}
```

Also, it isn't reliable which is why I think you had issues with migrations. I found that _Undo.RecordObject_ can be more reliable but neither are reliable on scene load.

There is no nice workaround unfortunately. _SetDirty_ would need to be called in _OnEnable_ at the earliest. Or in OnValidate. But both would need some flag to prevent _Undo.RecordObject_ from messing with undo/redo history and only trigger on migration.

Since this code is permanent, we could skip dirtying the scene? Although I called it a free migration it more saves us having to do a migration on our own data. But really it is just a normal code path.

Reply:
My usual thing is that any serialised data visible to the user should either be what is on disk, or dirty. Not doing this opens up edge cases where its auto picking Mesh but then i later add a Spline and then get confused why suddenly its picking Spline instead, stuff like that. But thats probably a 0.00001% edge case.

Hm disappointing to hear its unreliable, i feel like its been solid for me but i'm not sure i've called it from awake in the past.

In light of this i agree on foregoing it as the edge cases are tiny as noted above.

Reply:
Removed now

Reply:
> My usual thing is that any serialised data visible to the user should either be what is on disk, or dirty. Not doing this opens up edge cases where its auto picking Mesh but then i later add a Spline and then get confused why suddenly its picking Spline in my standalone builds when i thought it was set to Mesh, stuff like that. But thats probably a 0.00001% edge case.

Good point and agreed. In that scenario the added spline will dirty the scene for us.

> Hm disappointing to hear its unreliable, i feel like its been solid for me but i'm not sure i've called it from awake in the past. Or i just haven't noticed it failing..

The ODC may have been masking the issue when it creates the SFD input (or anything else that creates an ocean input from script) as this will trigger the SetDirty later.

Reply:
this was haunting me a bit so i went back to revist and you're right, setdirty is doing nothing here! i guess it doesnt work from "runtime" functions like `Awake()`. i've never had an issue calling it form editors etc. in this case i'm moving on, but i think this should be an exception to the rule - in general data should either match disk or be dirty. thanks again for raising this thread.